src/main/java/net/ripe/ipresource/scratch/AsnRange.java

```diff
@@ -0,0 +1,192 @@
+
+import net.ripe.ipresource.IpResourceType;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import static java.lang.Integer.toUnsignedLong;
+import static java.lang.Long.max;
+import static java.lang.Long.min;
+
+public final class AsnRange implements NumberResourceRange {
+    private final int start;
+    private final int end;
+
+    AsnRange(long start, long end) {
+        this.start = (int) start;
+        this.end = (int) end;
+        if (toUnsignedLong(this.start) != start) {
+            throw new IllegalArgumentException("start out of bounds");
+        }
+        if (toUnsignedLong(this.end) != end) {
+            throw new IllegalArgumentException("end out of bounds");
+        }
+        if (Integer.compareUnsigned(this.start, this.end) > 0) {
+            throw new IllegalArgumentException("start must be less than or equal to end");
+        }
+    }
+
+    private AsnRange(Asn start, Asn end) {
+        this(start.longValue(), end.longValue());
+    }
+
+    public static AsnRange range(Asn start, Asn end) {
+        return new AsnRange(start, end);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        return obj instanceof AsnRange that && this.start == that.start && this.end == that.end;
+    }
+
+    @Override
+    public int hashCode() {
+        return 'A' + 31 * 31 * Integer.hashCode(start) + 31 * Integer.hashCode(end);
+    }
+
+    @Override
+    public String toString() {
+        return isSingleton() ? String.valueOf(start()) : (start() + "-" + end());
+    }
+
+    @Override
+    public int compareTo(@NotNull NumberResourceRange o) {
+        return switch (o) {
+            case AsnRange that -> {
+                int rc = Integer.compareUnsigned(this.start, that.start);
+                if (rc != 0) {
+                    yield rc;
+                }
+                yield -Integer.compareUnsigned(this.end, that.end);
+            }
+            case IpBlock ignored -> -1;
+        };
+    }
+
+    @Override
+    public IpResourceType getType() {
+        return IpResourceType.ASN;
+    }
+
+    @Override
+    public Asn start() {
+        return new Asn(start);
+    }
+
+    @Override
+    public Asn end() {
+        return new Asn(end);
+    }
+
+    @Override
+    public boolean contains(@Nullable NumberResourceRange other) {
+        return switch (other) {
+            case null, IpBlock ignored -> false;
+            case AsnRange that -> this.lowerBound() <= that.lowerBound() && this.upperBound() >= that.upperBound();
+        };
+    }
+
+    @Override
+    public boolean isSingleton() {
+        return start == end;
+    }
+
+    @Override
+    public @NotNull List<@NotNull NumberResourceRange> subtract(@Nullable NumberResourceRange other) {
+        return switch (other) {
+            case null, IpBlock ignored -> Collections.emptyList();
```

Comment:
Interesting one, could also return `[this]`?

Reply:
True, is probably more correct indeed